pkg/controller/client_builder_dynamic.go

```diff
@@ -0,0 +1,222 @@
+		expirationSeconds:    defaultExpirationSeconds,
+		leewayPercent:        defaultLeewayPercent,
+		clock:                clock.RealClock{},
+	}
+	return builder
+}
+
+// this function only for test purpose, don't call it
+func NewTestDynamicClientBuilder(clientConfig *restclient.Config, coreClient v1core.CoreV1Interface, authClient v1authentication.AuthenticationV1Interface, ns string, expirationSeconds int64, leewayPercent int) ControllerClientBuilder {
+	builder := &DynamicControllerClientBuilder{
+		ClientConfig:         clientConfig,
+		CoreClient:           coreClient,
+		AuthenticationClient: authClient,
+		Namespace:            ns,
+		roundTripperFuncMap:  map[string]func(http.RoundTripper) http.RoundTripper{},
+		expirationSeconds:    expirationSeconds,
+		leewayPercent:        leewayPercent,
+		clock:                clock.RealClock{},
+	}
+	return builder
+}
+
+func (t *DynamicControllerClientBuilder) Config(saName string) (*restclient.Config, error) {
+	_, err := getOrCreateServiceAccount(t.CoreClient, t.Namespace, saName)
+	if err != nil {
+		return nil, err
+	}
+
+	configCopy := constructClient(t.Namespace, saName, t.ClientConfig)
+
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	rt, ok := t.roundTripperFuncMap[saName]
+	if ok {
+		configCopy.WrapTransport = rt
+	} else {
+		cachedTokenSource := transport.NewCachedTokenSource(&tokenSourceImpl{
+			clientConfig:       t.ClientConfig,
+			namespace:          t.Namespace,
+			serviceAccountName: saName,
+			coreClient:         t.CoreClient,
+			authClient:         t.AuthenticationClient,
+			expirationSeconds:  t.expirationSeconds,
+			leewayPercent:      t.leewayPercent,
+		})
+		configCopy.WrapTransport = transport.TokenSourceWrapTransport(cachedTokenSource)
+
+		t.roundTripperFuncMap[saName] = configCopy.WrapTransport
+	}
+
+	return &configCopy, nil
+}
+
+func (t *DynamicControllerClientBuilder) ConfigOrDie(name string) *restclient.Config {
+	clientConfig, err := t.Config(name)
+	if err != nil {
+		klog.Fatal(err)
+	}
+	return clientConfig
+}
+
+func (t *DynamicControllerClientBuilder) Client(name string) (clientset.Interface, error) {
+	clientConfig, err := t.Config(name)
+	if err != nil {
+		return nil, err
+	}
+	return clientset.NewForConfig(clientConfig)
+}
+
+func (t *DynamicControllerClientBuilder) ClientOrDie(name string) clientset.Interface {
+	client, err := t.Client(name)
+	if err != nil {
+		klog.Fatal(err)
+	}
+	return client
+}
+
+type tokenSourceImpl struct {
+	clientConfig       *restclient.Config
+	namespace          string
+	serviceAccountName string
+	coreClient         v1core.CoreV1Interface
+	authClient         v1authentication.AuthenticationV1Interface
+	expirationSeconds  int64
+	leewayPercent      int
+}
+
+func (ts *tokenSourceImpl) Token() (*oauth2.Token, error) {
+	var retTokenRequest *v1authenticationapi.TokenRequest
+
+	backoff := wait.Backoff{
+		Duration: 5 * time.Second,
+		Factor:   2, // double the timeout for every failure
+		Steps:    10,
```

Comment:
this number of steps means we'd wait ~1.5 hours before returning failure, and the last attempt would pause for ~40 minutes... that seems excessive for an inner loop. I'd actually expect the client timeouts to have kicked in and failed the request after about 30 seconds or so.

I'd suggest starting this number way lower (~500ms) and having fewer steps (~4), letting this inner loop take up to ~7 seconds before returning